File: leetcode/54/spiralMatrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m = len(matrix)
        n = len(matrix[0])
        visited = [[False for _ in range(n)] for _ in range(m)] # 방문 여부 확인
        dx = [1, 0, -1, 0]
        dy = [0, 1, 0, -1]
        d = 0 # 방향 설정
        x = 0
        y = 0
        tmp = 0
        visited[0][0] = True

        answer = [matrix[0][0]]

        while tmp < 2:
            xx = x+dx[d]
            yy = y+dy[d]
            if 0 <= xx < n and 0 <= yy < m and visited[yy][xx] == False:
                x = xx
                y = yy
                visited[y][x] = True
                answer.append(matrix[y][x])
                tmp = 0
            else:
                d = (d+1)%4 # 진행 방향이 방문된 곳이면 방향 변경
                tmp += 1
            

        # print(visited)
        return answer

File: leetcode/54/spiralMatrix.swift
class Solution {
    func spiralOrder(_ matrix: [[Int]]) -> [Int] {
        let m = matrix.count
        let n = matrix[0].count
        var visited = Array(repeating:Array(repeating:false, count:n), count: m)
        let dr = [0, 1, 0, -1]
        let dc = [1, 0, -1, 0]
        var d = 0
        var r = 0
        var c = 0
        var tmp = 0
        var answer = [matrix[0][0]]
        visited[0][0] = true

        while tmp < 2{
            // print(visited, r, c, d)
            let rr = r + dr[d]
            let cc = c + dc[d]
            if rr >= 0 && rr < m && cc >= 0 && cc < n && !visited[rr][cc] {
                r = rr
                c = cc
                visited[r][c] = true
                answer.append(matrix[r][c])
                tmp = 0
            }
            else{
                d = (d+1)%4
                tmp += 1
            }
        }
        return answer
    }
}
